File: Mba.Simplifier/Jit/FastAmd64Assembler.cs
using Iced.Intel;
using Mba.Common.MSiMBA;
using Mba.Simplifier.Interpreter;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics.Arm;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace Mba.Simplifier.Jit
{
    unsafe ref struct StackBuffer
    {
        public byte* Ptr;

        public uint Offset;

        public StackBuffer(byte* ptr)
        {
            this.Ptr = ptr;
        }

        public void PushU8(byte value)
        {
            Ptr[Offset++] = value;
        }

        public void PushI32(int value)
            => PushU32((uint)value);

        public void PushU32(uint value)
        {
            *(uint*)&Ptr[Offset] = value;
            Offset += 4;
        }

        public void PushU64(ulong value)
        {
            *(ulong*)&Ptr[Offset] = value;
            Offset += 8;
        }
    }

    public unsafe class FastAmd64Assembler : IAmd64Assembler
    {
        public byte* start;

        public byte* ptr;

        public int offset = 0;

        public List<Instruction> Instructions => GetInstructions();

        public FastAmd64Assembler(byte* ptr)
        {
            start = ptr;
            this.ptr = ptr;
        }

        private unsafe void EmitBytes(params byte[] bytes)
        {
            fixed(byte* p = &bytes[0])
            {
                Memcpy(ptr, p, (uint)bytes.Length);
            }

            ptr += bytes.Length;
        }

        private void EmitBuffer(StackBuffer buffer)
        {
            Memcpy(ptr, buffer.Ptr, buffer.Offset);
            ptr += buffer.Offset;
        }

        private void Memcpy(void* destination, void* source, uint byteCount)
        {
            Unsafe.CopyBlockUnaligned(destination, source, byteCount);
        }

        public void PushReg(Register reg)
        {
            if (reg >= Register.RAX && reg <= Register.RDI)
            {
                byte opcode = (byte)(0x50 + GetRegisterCode(reg));
                EmitBytes(opcode);
                return;
            }

            if (reg >= Register.R8 && reg <= Register.R15)
            {
                byte rex = 0x41;
                byte opcode = (byte)(0x50 + (int)reg - (int)Register.R8);
                EmitBytes(rex, opcode);
                return;
            }

            throw new ArgumentException("Invalid register for PUSH instruction.");
            
        }
        // push qword ptr [baseReg+offset]
        public void PushMem64(Register baseReg, int offset)
        {
            byte* p = stackalloc byte[8];
            var arr = new StackBuffer(ptr);

            if (IsExtended(baseReg))
            {
                byte rex = 0x49;
                arr.PushU8(rex);
            }

            byte opcode = 0xFF;
            byte modrm = (byte)(0x80 | (0x06 << 3) | (GetRegisterCode(baseReg) & 0x07));
            arr.PushU8(opcode);
            arr.PushU8(modrm);

            if (baseReg == Register.RSP || baseReg == Register.R12)
            {
                byte sib = (byte)(0x00 | (0x04 << 3) | (GetRegisterCode(baseReg) & 0x07));
                arr.PushU8(sib); 
            }

            arr.PushI32(offset);

            EmitBuffer(arr);
        }

        public void PopReg(Register reg)
        {
            if (reg >= Register.RAX && reg <= Register.RDI)
            {
                byte opcode = (byte)(0x58 + GetRegisterCode(reg));
                EmitBytes(opcode);
                return;
            }

            if (reg >= Register.R8 && reg <= Register.R15)
            {
                byte rex = 0x41;
                byte opcode = (byte)(0x58 + GetRegisterCode(reg) - 8);
                EmitBytes(rex, opcode);
                return;
            }

            throw new ArgumentException($"Cannot pop {reg}");
            
        }

        public void OpcodeRegReg(byte opcode, Register reg1, Register reg2)
        {
            byte rex = 0x48; 
            if (IsExtended(reg1)) rex |= 0x01; 
            if (IsExtended(reg2)) rex |= 0x04;

            byte modrm = (byte)(0xC0 | ((GetRegisterCode(reg2) & 0x07) << 3) | (GetRegisterCode(reg1) & 0x07));
            EmitBytes(rex, opcode, modrm);
        }

        public void MovRegReg(Register reg1, Register reg2)
            => OpcodeRegReg(0x89, reg1, reg2);

        public void MovRegMem64(Register dstReg, Register baseReg, int offset)
        {
            byte* p = stackalloc byte[8];
            var arr = new StackBuffer(ptr);

            byte rex = 0x48; 
            if (IsExtended(dstReg)) rex |= 0x04;
            if (IsExtended(baseReg)) rex |= 0x01; 

            byte opcode = 0x8B;
            byte modrm = (byte)(0x80 | ((GetRegisterCode(dstReg) & 0x07) << 3) | (GetRegisterCode(baseReg) & 0x07));

            arr.PushU8(rex);
            arr.PushU8(opcode);
            arr.PushU8(modrm);

            if (baseReg == Register.RSP || baseReg == Register.R12)
            {
                byte sib = (byte)(0x00 | (0x04 << 3) | (GetRegisterCode(baseReg) & 0x07));
                arr.PushU8(sib);
            }

            arr.PushI32(offset);

            EmitBuffer(arr);
        }

        // mov qword ptr [baseReg + offset], srcReg
        public void MovMem64Reg(Register baseReg, int offset, Register srcReg)
        {
            byte* p = stackalloc byte[8];
            var arr = new StackBuffer(ptr);

            byte rex = 0x48; 
            if (IsExtended(srcReg)) rex |= 0x04; 
            if (IsExtended(baseReg)) rex |= 0x01; 

            byte opcode = 0x89;
            byte modrm = (byte)(0x80 | ((GetRegisterCode(srcReg) & 0x07) << 3) | (GetRegisterCode(baseReg) & 0x07));

            arr.PushU8(rex);
            arr.PushU8(opcode);
            arr.PushU8(modrm);

            if (baseReg == Register.RSP || baseReg == Register.R12)
            {
                byte sib = (byte)(0x00 | (0x04 << 3) | (GetRegisterCode(baseReg) & 0x07));
                arr.PushU8(sib);
            }

            arr.PushI32(offset);
            EmitBuffer(arr);
        }

        public void MovabsRegImm64(Register reg1, ulong imm64)
        {
            byte rex = 0x48;
            if (IsExtended(reg1)) rex |= 0x01;

            var cond = (reg1 >= Register.RAX && reg1 <= Register.RDI);
            byte opcode = (byte)(0xB8 + (cond ? GetRegisterCode(reg1) : GetRegisterCode(reg1) - 8));

            byte* p = stackalloc byte[10];
            var arr = new StackBuffer(ptr);
            arr.PushU8(rex);
            arr.PushU8(opcode);
            arr.PushU64(imm64);

            EmitBuffer(arr);
        }

        public void AddRegReg(Register reg1, Register reg2)
        {
            OpcodeRegReg(0x01, reg1, reg2);
        }

        public void AddRegImm32(Register reg1, uint imm32)
            => OpcRegImm(0x00, reg1, imm32);

        public void SubRegImm32(Register reg1, uint imm32)
            => OpcRegImm(0x05, reg1, imm32);

        public void OpcRegImm(byte mask, Register reg1, uint imm32)
        {
            byte rex = 0x48;
            if (IsExtended(reg1)) rex |= 0x01;

            byte opcode = 0x81;
            byte modrm = (byte)(0xC0 | (mask << 3) | (GetRegisterCode(reg1) & 0x07));

            byte* p = stackalloc byte[7];
            var arr = new StackBuffer(ptr);
            arr.PushU8(rex);
            arr.PushU8(opcode);
            arr.PushU8(modrm);
            arr.PushU32(imm32);

            EmitBuffer(arr);
        }

        public void ImulRegReg(Register reg1, Register reg2)
        {
            byte rex = 0x48;
            if (IsExtended(reg1)) rex |= 0x04;
            if (IsExtended(reg2)) rex |= 0x01;

            byte opcode1 = 0x0F;
            byte opcode2 = 0xAF;
            byte modrm = (byte)(0xC0 | ((GetRegisterCode(reg1) & 0x07) << 3) | (GetRegisterCode(reg2) & 0x07));
            EmitBytes(rex, opcode1, opcode2, modrm);
        }

        public void AndRegReg(Register reg1, Register reg2)
            => OpcodeRegReg(0x21, reg1, reg2);

        public void AndRegImm32(Register reg1, uint imm32)
            => OpcRegImm(0x04, reg1, imm32);

        public void AndMem64Reg(Register baseReg, int offset, Register srcReg)
        {
            byte rex = 0x48; 
            if (IsExtended(srcReg)) rex |= 0x04; 
            if (IsExtended(baseReg)) rex |= 0x01;

            byte opcode = 0x21;
            byte modrm = (byte)(0x80 | ((GetRegisterCode(srcReg) & 0x07) << 3) | (GetRegisterCode(baseReg) & 0x07));

            byte* p = stackalloc byte[8];
            var arr = new StackBuffer(ptr);
            arr.PushU8(rex);
            arr.PushU8(opcode);
            arr.PushU8(modrm);


            if (baseReg == Register.RSP || baseReg == Register.R12)
            {
                byte sib = (byte)(0x00 | (0x04 << 3) | (GetRegisterCode(baseReg) & 0x07));
                arr.PushU8(sib);
            }

            arr.PushI32(offset);

            EmitBuffer(arr);
        }

        public void OrRegReg(Register reg1, Register reg2)
            => OpcodeRegReg(0x09, reg1, reg2);

        public void XorRegReg(Register reg1, Register reg2)
            => OpcodeRegReg(0x31, reg1, reg2);

        public void NotReg(Register reg1)
        {
            byte rex = 0x48;
            if (IsExtended(reg1)) rex |= 0x01;

            byte opcode = 0xF7;
            byte modrm = (byte)(0xC0 | (0x02 << 3) | (GetRegisterCode(reg1) & 0x07));
            EmitBytes(rex, opcode, modrm);
        }

        public void ShlRegCl(Register reg)
            => ShiftRegCl(true, reg);

        public void ShrRegCl(Register reg)
            => ShiftRegCl(false, reg);

        public void ShiftRegCl(bool shl, Register reg)
        {
            byte rex = 0x48;
            if (IsExtended(reg)) rex |= 0x01;

            byte opcode = 0xD3;
            var m1 = shl ? 0x04 : 0x05;
            byte modrm = (byte)(0xC0 | (m1 << 3) | (GetRegisterCode(reg) & 0x07));
            EmitBytes(rex, opcode, modrm);
        }

        public void ShrRegImm8(Register reg, byte imm8)
        {
            byte rex = 0x48;
            if (IsExtended(reg)) rex |= 0x01;

            byte opcode = 0xC1;
            byte modrm = (byte)(0xC0 | (0x05 << 3) | (GetRegisterCode(reg) & 0x07));
            EmitBytes(rex, opcode, modrm, imm8);
        }

        public void CallReg(Register reg1)
        {
            byte rex = 0x00;
            if (IsExtended(reg1))
                rex = 0x41;

            byte opcode = 0xFF;
            byte modrm = (byte)(0xC0 | (0x02 << 3) | (GetRegisterCode(reg1) & 0x07));
            if (rex != 0)
                EmitBytes(rex, opcode, modrm);
            else
                EmitBytes(opcode, modrm);
        }

        public void Ret()
            => EmitBytes(0xC3);

        private bool IsExtended(Register reg)
            => reg >= Register.R8 && reg <= Register.R15;

        private uint GetRegisterCode(Register reg)
            => (uint)reg - (uint)Register.RAX;

        public List<Instruction> GetInstructions()
        {
            var ourBytes = GetBytes().ToArray();
            var codeReader = new ByteArrayCodeReader(ourBytes);
            var decoder = Iced.Intel.Decoder.Create(64, codeReader);
            decoder.IP = 0;

            var insts = new List<Instruction>();
            while (true)
            {
                insts.Add(decoder.Decode());
                if (codeReader.Position >= ourBytes.Length)
                    break;
            }

            return insts;
        }

        public List<byte> GetBytes()
        {
            var len = (ulong)ptr - (ulong)start;
            var output = new List<byte>();
            for (ulong i = 0; i < len; i++)
                output.Add(start[i]);

            return output;
        }

        public void Reset()
            => ptr = start;
    }

    public static class Reg
    {
        public static AssemblerRegister64 Conv(Register reg)
        {
            return reg switch
            {
                Register.RCX => AssemblerRegisters.rcx,
                Register.RDX => AssemblerRegisters.rdx,
                Register.RBX => AssemblerRegisters.rbx,
                Register.RDI => AssemblerRegisters.rdi,
                Register.RAX => AssemblerRegisters.rax,
                Register.RSI => AssemblerRegisters.rsi,
                Register.R8 => AssemblerRegisters.r8,
                Register.R9 => AssemblerRegisters.r9,
                Register.R10 => AssemblerRegisters.r10,
                Register.R11 => AssemblerRegisters.r11,
                Register.R12 => AssemblerRegisters.r12,
                Register.R13 => AssemblerRegisters.r13,
                Register.R14 => AssemblerRegisters.r14,
                Register.R15 => AssemblerRegisters.r15,
                Register.RBP => AssemblerRegisters.rbp,
                Register.RSP => AssemblerRegisters.rsp,
                _ => throw new InvalidOperationException()
            };
        }
    }
}

File: EqSat/src/assembler/fast_amd64_assembler.rs
use iced_x86::code_asm::*;
use iced_x86::{Instruction, Register};
use rand::Rng;
use std::fmt::Write;
use std::hint::unreachable_unchecked;
use std::time::Instant;

use crate::assembler::amd64_assembler::IAmd64Assembler;

// Wrapper around a stack-allocated byte buffer for building instruction byte sequences
pub struct StackBuffer<'a> {
    pub arr: &'a mut [u8],
    pub offset: usize,
}

impl StackBuffer<'_> {
    fn push<T>(&mut self, v: T) {
        unsafe {
            let ptr = self.arr.as_mut_ptr().add(self.offset) as *mut T;
            *ptr = v;
        }

        self.offset += std::mem::size_of::<T>();
    }

    fn pop<T: Copy>(&mut self) -> T {
        self.offset -= std::mem::size_of::<T>();

        unsafe {
            let ptr = self.arr.as_mut_ptr().add(self.offset) as *mut T;
            *ptr
        }
    }

    fn push_u8(&mut self, byte: u8) {
        unsafe {
            *self.arr.get_unchecked_mut(self.offset) = byte;
        }

        self.offset += 1;
    }

    fn push_i32(&mut self, byte: i32) {
        unsafe {
            let ptr = self.arr.as_mut_ptr().add(self.offset) as *mut i32;
            *ptr = byte;
        }

        self.offset += 4;
    }

    fn push_u32(&mut self, byte: u32) {
        unsafe {
            let ptr = self.arr.as_mut_ptr().add(self.offset) as *mut u32;
            *ptr = byte;
        }

        self.offset += 4;
    }

    fn push_u64(&mut self, byte: u64) {
        unsafe {
            let ptr = self.arr.as_mut_ptr().add(self.offset) as *mut u64;
            *ptr = byte;
        }

        self.offset += 8;
    }
}

pub struct FastAmd64Assembler {
    pub p: *mut u8,
    pub offset: usize,
}

impl FastAmd64Assembler {
    pub fn new(buffer: *mut u8) -> Self {
        FastAmd64Assembler {
            p: buffer,
            offset: 0,
        }
    }

    fn emit_bytes(&mut self, data: &[u8]) {
        unsafe {
            std::ptr::copy_nonoverlapping(data.as_ptr(), self.p.add(self.offset), data.len());
        }

        self.offset += data.len();
    }

    fn emit_buffer(&mut self, buffer: &StackBuffer) {
        unsafe {
            std::ptr::copy_nonoverlapping(
                buffer.arr.as_ptr(),
                self.p.add(self.offset),
                buffer.offset,
            );
        }

        self.offset += buffer.offset;
    }

    pub fn opcode_reg_reg(&mut self, opcode: u8, reg1: Register, reg2: Register) {
        let mut rex = 0x48;
        if is_extended(reg1) {
            rex |= 0x01;
        }
        if is_extended(reg2) {
            rex |= 0x04;
        }

        let modrm = 0xC0
            | ((get_register_code(reg2) as u8 & 0x07) << 3)
            | (get_register_code(reg1) as u8 & 0x07);
        self.emit_bytes(&[rex, opcode, modrm]);
    }

    pub fn opc_reg_imm(&mut self, mask: u8, reg: Register, imm32: u32) {
        let p = &mut [0u8; 7];
        let mut arr = StackBuffer { arr: p, offset: 0 };

        let mut rex = 0x48;
        if self.is_extended(reg) {
            rex |= 0x01;
        }

        let opcode = 0x81;
        let modrm = 0xC0 | (mask << 3) | (self.get_register_code(reg) & 0x07);

        arr.push_u8(rex);
        arr.push_u8(opcode);
        arr.push_u8(modrm);
        arr.push_u32(imm32);

        self.emit_buffer(&arr);
    }

    pub fn shift_reg_cl(&mut self, shl: bool, reg: Register) {
        let mut rex = 0x48;
        if self.is_extended(reg) {
            rex |= 0x01;
        }

        let opcode = 0xD3;
        let m1 = if shl { 0x04 } else { 0x05 };
        let modrm = 0xC0 | (m1 << 3) | (self.get_register_code(reg) & 0x07);

        self.emit_bytes(&[rex, opcode, modrm]);
    }

    #[inline(always)]
    fn mov_reg_mem64_template(&mut self, dst_reg: Register, base_reg: Register, offset: i32) {
        let p = &mut [0u8; 8];
        let mut arr = StackBuffer { arr: p, offset: 0 };

        let mut rex = 0x48;
        if is_extended(dst_reg) {
            rex |= 0x04;
        }
        if is_extended(base_reg) {
            rex |= 0x01;
        }

        let opcode = 0x8B;
        let modrm = 0x80
            | ((get_register_code(dst_reg) as u8 & 0x07) << 3)
            | (get_register_code(base_reg) as u8 & 0x07);

        arr.push_u8(rex);
        arr.push_u8(opcode);
        arr.push_u8(modrm);

        if base_reg == Register::RSP || base_reg == Register::R12 {
            let sib = 0x00 | (0x04 << 3) | (get_register_code(base_reg) as u8 & 0x07);
            arr.push_u8(sib);
        }

        arr.push_i32(offset);

        self.emit_buffer(&arr);
    }

    pub fn is_extended(&mut self, reg: Register) -> bool {
        return reg >= Register::R8 && reg <= Register::R15;
    }

    pub fn get_register_code(&mut self, reg: Register) -> u8 {
        return (reg as u8) - (Register::RAX as u8);
    }
}

impl IAmd64Assembler for FastAmd64Assembler {
    fn push_reg(&mut self, reg: Register) {
        if reg >= Register::RAX && reg <= Register::RDI {
            let opcode = (0x50 + get_register_code(reg)) as u8;
            self.emit_bytes(&[opcode]);
            return;
        }

        let rex = 0x41;
        let opcode = (0x50 + reg as u8 - Register::R8 as u8);
        self.emit_bytes(&[rex, opcode]);
    }

    fn push_mem64(&mut self, base_reg: Register, disp: i32) {
        let p = &mut [0u8; 8];
        let mut arr = StackBuffer { arr: p, offset: 0 };

        if is_extended(base_reg) {
            let rex = 0x49;
            arr.push_u8(rex);
        }

        let opcode: u8 = 0xFF;
        let modrm = (0x80 | (0x06 << 3) | (get_register_code(base_reg) & 0x07)) as u8;
        arr.push_u8(opcode);
        arr.push_u8(modrm);

        if base_reg == Register::RSP || base_reg == Register::R12 {
            let sib: u8 = (0x00 | (0x04 << 3) | (get_register_code(base_reg) & 0x07)) as u8;
            arr.push_u8(sib);
        }

        arr.push_i32(disp);
        self.emit_buffer(&arr);
    }

    fn pop_reg(&mut self, reg: Register) {
        if reg >= Register::RAX && reg <= Register::RDI {
            let opcode = 0x58 + get_register_code(reg) as u8;
            self.emit_bytes(&[opcode]);
            return;
        }

        let rex = 0x41;
        let opcode = 0x58 + get_register_code(reg) as u8 - 8;
        self.emit_bytes(&[rex, opcode]);
        return;
    }

    fn mov_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.opcode_reg_reg(0x89, reg1, reg2);
    }

    fn mov_reg_mem64(&mut self, dst_reg: Register, base_reg: Register, offset: i32) {
        // This function is intentionally monomorphized for performance.
        // mov_reg_mem64 has a variable length encoding depending on the base register,
        // which prevents SROA from promoting our stack buffer to an SSA variable.
        if base_reg == Register::R12 {
            self.mov_reg_mem64_template(dst_reg, Register::R12, offset);
        } else if base_reg == Register::RSP {
            self.mov_reg_mem64_template(dst_reg, Register::RSP, offset);
        } else {
            self.mov_reg_mem64_template(dst_reg, base_reg, offset);
        }
    }

    fn mov_mem64_reg(&mut self, base_reg: Register, offset: i32, src_reg: Register) {
        let p = &mut [0u8; 8];
        let mut arr = StackBuffer { arr: p, offset: 0 };

        let mut rex = 0x48;
        if self.is_extended(src_reg) {
            rex |= 0x04;
        }
        if self.is_extended(base_reg) {
            rex |= 0x01;
        }

        let opcode = 0x89;
        let modrm = 0x80
            | ((self.get_register_code(src_reg) & 0x07) << 3)
            | (self.get_register_code(base_reg) & 0x07);

        arr.push_u8(rex);
        arr.push_u8(opcode);
        arr.push_u8(modrm);

        if base_reg == Register::RSP || base_reg == Register::R12 {
            let sib = 0x00 | (0x04 << 3) | (self.get_register_code(base_reg) & 0x07);
            arr.push_u8(sib);
        }

        arr.push_i32(offset);

        self.emit_buffer(&arr);
    }

    fn movabs_reg_imm64(&mut self, reg: Register, imm64: u64) {
        let p = &mut [0u8; 10];
        let mut arr = StackBuffer { arr: p, offset: 0 };

        let mut rex = 0x48;
        if self.is_extended(reg) {
            rex |= 0x01;
        }

        let cond = reg >= Register::RAX && reg <= Register::RDI;
        let opcode = 0xB8
            + if cond {
                self.get_register_code(reg)
            } else {
                self.get_register_code(reg) - 8
            };

        arr.push_u8(rex);
        arr.push_u8(opcode);
        arr.push_u64(imm64);

        self.emit_buffer(&arr);
    }

    fn add_reg_reg(&mut self, dest: Register, src: Register) {
        self.opcode_reg_reg(0x01, dest, src);
    }

    fn add_reg_imm32(&mut self, reg: Register, imm32: u32) {
        self.opc_reg_imm(0x00, reg, imm32);
    }

    fn sub_reg_imm32(&mut self, reg: Register, imm32: u32) {
        self.opc_reg_imm(0x05, reg, imm32);
    }

    fn imul_reg_reg(&mut self, reg1: Register, reg2: Register) {
        let mut rex = 0x48;
        if self.is_extended(reg1) {
            rex |= 0x04;
        }
        if self.is_extended(reg2) {
            rex |= 0x01;
        }

        let opcode1 = 0x0F;
        let opcode2 = 0xAF;
        let modrm = 0xC0
            | ((self.get_register_code(reg1) & 0x07) << 3)
            | (self.get_register_code(reg2) & 0x07);

        self.emit_bytes(&[rex, opcode1, opcode2, modrm]);
    }

    fn and_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.opcode_reg_reg(0x21, reg1, reg2);
    }

    fn and_reg_imm32(&mut self, reg1: Register, imm32: u32) {
        self.opc_reg_imm(0x04, reg1, imm32);
    }

    fn and_mem64_reg(&mut self, base_reg: Register, offset: i32, src_reg: Register) {
        let p = &mut [0u8; 8];
        let mut arr = StackBuffer { arr: p, offset: 0 };

        let mut rex = 0x48;
        if self.is_extended(src_reg) {
            rex |= 0x04;
        }
        if self.is_extended(base_reg) {
            rex |= 0x01;
        }

        let opcode = 0x21;
        let modrm = 0x80
            | ((self.get_register_code(src_reg) & 0x07) << 3)
            | (self.get_register_code(base_reg) & 0x07);

        arr.push_u8(rex);
        arr.push_u8(opcode);
        arr.push_u8(modrm);

        if base_reg == Register::RSP || base_reg == Register::R12 {
            let sib = 0x00 | (0x04 << 3) | (self.get_register_code(base_reg) & 0x07);
            arr.push_u8(sib);
        }

        arr.push_i32(offset);

        self.emit_buffer(&arr);
    }

    fn or_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.opcode_reg_reg(0x09, reg1, reg2);
    }

    fn xor_reg_reg(&mut self, reg1: Register, reg2: Register) {
        self.opcode_reg_reg(0x31, reg1, reg2);
    }

    fn not_reg(&mut self, reg: Register) {
        let mut rex = 0x48;
        if self.is_extended(reg) {
            rex |= 0x01;
        }

        let opcode = 0xF7;
        let modrm = 0xC0 | (0x02 << 3) | (self.get_register_code(reg) & 0x07);

        self.emit_bytes(&[rex, opcode, modrm]);
    }

    fn shl_reg_cl(&mut self, reg: Register) {
        self.shift_reg_cl(true, reg);
    }

    fn shr_reg_cl(&mut self, reg: Register) {
        self.shift_reg_cl(false, reg);
    }

    fn shr_reg_imm8(&mut self, reg: Register, imm8: u8) {
        let mut rex = 0x48;
        if self.is_extended(reg) {
            rex |= 0x01;
        }

        let opcode = 0xC1;
        let modrm = 0xC0 | (0x05 << 3) | (self.get_register_code(reg) & 0x07);

        self.emit_bytes(&[rex, opcode, modrm, imm8]);
    }

    fn call_reg(&mut self, reg: Register) {
        let mut rex = 0x00;
        if self.is_extended(reg) {
            rex = 0x41;
        }

        let opcode = 0xFF;
        let modrm = 0xC0 | (0x02 << 3) | (self.get_register_code(reg) & 0x07);

        if rex != 0 {
            self.emit_bytes(&[rex, opcode, modrm]);
        } else {
            self.emit_bytes(&[opcode, modrm]);
        }
    }

    fn ret(&mut self) {
        self.emit_bytes(&[0xC3]);
    }

    fn get_instructions(&mut self) -> Vec<Instruction> {
        let bytes = self.get_bytes();
        let mut decoder = iced_x86::Decoder::new(64, &bytes, iced_x86::DecoderOptions::NONE);
        decoder.set_ip(0);

        let mut instructions = Vec::new();
        while decoder.position() < bytes.len() {
            let instruction = decoder.decode();
            instructions.push(instruction);
        }

        instructions
    }

    fn get_bytes(&mut self) -> Vec<u8> {
        let mut bytes = Vec::new();
        for i in 0..self.offset {
            unsafe {
                bytes.push(*self.p.add(i));
            }
        }
        return bytes;
    }

    fn reset(&mut self) {
        self.offset = 0;
    }
}

fn is_extended(reg: Register) -> bool {
    return reg >= Register::R8 && reg <= Register::R15;
}

fn get_register_code(reg: Register) -> u8 {
    return (reg as u8) - (Register::RAX as u8);
}
